<div dir="rtl" align="center" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; background: linear-gradient(135deg, #4a148c 0%, #1565c0 55%, #00897b 100%); color: white; padding: 28px 20px; border-radius: 18px; margin-bottom: 10px;">

<p style="font-size: 15px; margin: 0; opacity: 0.9;">🐍 دوره‌ی «ماجراجویی با پایتون» — ویژه‌ی نوجوانان</p>

# جلسه‌ی ۲: حلقه‌ی for روی لیست

<p style="font-size: 17px; margin: 6px 0;">به کامپیوتر یاد می‌دهیم که یک کار را بارها و بارها، بدون خستگی تکرار کند.</p>

<p style="font-size: 14px; margin: 0; opacity: 0.9;">⏱️ زمان پیشنهادی: ۹۰ دقیقه</p>

</div>

<div dir="rtl" align="center" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif;">

🟪🟪⬜⬜⬜⬜⬜⬜⬜⬜⬜ &nbsp; جلسه‌ی ۲ از ۱۱

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #e8eaf6; color: #1b1b1b; border-right: 6px solid #3949ab; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 🎯 اهداف این جلسه

در پایان این جلسه می‌توانی:

1. با حلقه‌ی <code dir="ltr">for</code> روی **همه‌ی اعضای یک لیست** حرکت کنی.
2. نقش **دونقطه** <code dir="ltr">:</code> و **تورفتگی** (indentation) را در پایتون توضیح بدهی.
3. با الگوی **جمع‌کننده** (accumulator) مجموع و میانگین یک لیست را حساب کنی.
4. با الگوی **شمارنده** تعداد چیزی را بشماری.
5. روی حروف یک **رشته** حلقه بزنی.
6. با جدول **ردیابی** (trace table) دنبال کنی که حلقه در هر دور چه کار می‌کند.

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #e0f7fa; color: #1b1b1b; border-right: 6px solid #00838f; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 🔁 مرور جلسه‌ی قبل

- لیست را با <code dir="ltr">[ ]</code> می‌سازیم: <code dir="ltr">channels = ["Fp1", "C3", "O1"]</code>
- اندیس از **صفر** شروع می‌شود: <code dir="ltr">channels[0]</code> یعنی <code dir="ltr">"Fp1"</code>.
- <code dir="ltr">channels[-1]</code> آخرین عضو است و <code dir="ltr">len(channels)</code> تعداد اعضا را می‌دهد.

❓ **سؤال گرم‌کردن:** <code dir="ltr">channels[1:]</code> چه چیزی را برمی‌گرداند؟

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

<details>
<summary><b>👀 دیدن پاسخ</b></summary>

<code dir="ltr">['C3', 'O1']</code>؛ یعنی همه‌ی اعضا به جز اولی.

</details>

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

## ۱. مشکل: کارهای تکراری

فرض کن می‌خواهیم نام هر کانال EEG را با یک پیام چاپ کنیم. بدون حلقه باید این‌طور بنویسیم:

</div>

In [1]:
channels = ["Fp1", "Fp2", "C3", "C4", "O1", "O2"]

print("در حال بررسی کانال", channels[0])
print("در حال بررسی کانال", channels[1])
print("در حال بررسی کانال", channels[2])
print("در حال بررسی کانال", channels[3])
print("در حال بررسی کانال", channels[4])
print("در حال بررسی کانال", channels[5])

در حال بررسی کانال Fp1
در حال بررسی کانال Fp2
در حال بررسی کانال C3
در حال بررسی کانال C4
در حال بررسی کانال O1
در حال بررسی کانال O2


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

اگر ۶۴ کانال داشتیم چه؟ یا اگر تعداد کانال‌ها را از قبل نمی‌دانستیم؟ 😩

برنامه‌نویس‌ها یک قانون طلایی دارند: **«خودت را تکرار نکن!»** (به انگلیسی: Don't Repeat Yourself یا به اختصار DRY). برای کارهای تکراری از **حلقه** استفاده می‌کنیم.

## ۲. ساختار حلقه‌ی <code dir="ltr">for</code>

<div dir="ltr" align="left">

```python
for item in my_list:
    # کارهایی که برای هر عضو انجام می‌شود
```

</div>

این را این‌طور بخوان: **«برای هر <code dir="ltr">item</code> در <code dir="ltr">my_list</code>، کارهای زیر را انجام بده.»**

- <code dir="ltr">for</code> و <code dir="ltr">in</code> کلمه‌های کلیدی پایتون هستند.
- <code dir="ltr">item</code> یک **متغیر حلقه** است. در هر دور، یکی از اعضای لیست در آن قرار می‌گیرد. اسمش را خودمان انتخاب می‌کنیم.
- **دونقطه** <code dir="ltr">:</code> در انتهای خط اول اجباری است.
- خط‌هایی که با **تورفتگی** (معمولاً ۴ فاصله) نوشته شده‌اند، **بدنه‌ی حلقه** هستند و در هر دور اجرا می‌شوند.

</div>

In [2]:
channels = ["Fp1", "Fp2", "C3", "C4", "O1", "O2"]

for channel in channels:
    print("در حال بررسی کانال", channel)

print("✅ بررسی همه‌ی کانال‌ها تمام شد.")

در حال بررسی کانال Fp1
در حال بررسی کانال Fp2
در حال بررسی کانال C3
در حال بررسی کانال C4
در حال بررسی کانال O1
در حال بررسی کانال O2
✅ بررسی همه‌ی کانال‌ها تمام شد.


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #e8f5e9; color: #1b1b1b; border-right: 6px solid #2e7d32; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 💡 تورفتگی در پایتون معنا دارد

خط آخر برنامه‌ی بالا (<code dir="ltr">print("✅ ...")</code>) تورفتگی **ندارد**؛ پس جزء حلقه نیست و فقط **یک بار**، بعد از تمام شدن حلقه اجرا می‌شود.

در بسیاری از زبان‌های دیگر، تورفتگی فقط برای زیبایی است؛ اما در پایتون **بخشی از دستور زبان** است و تعیین می‌کند کدام خط‌ها داخل حلقه هستند.

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

## ۳. حلقه دقیقاً چه کار می‌کند؟ جدول ردیابی

بیا قدم‌به‌قدم ببینیم در هر دور حلقه چه اتفاقی می‌افتد. لیست سیاره‌های سنگی را در نظر بگیر:

</div>

In [3]:
rocky_planets = ["عطارد", "زهره", "زمین", "مریخ"]

for planet in rocky_planets:
    print("🪐", planet)

🪐 عطارد
🪐 زهره
🪐 زمین
🪐 مریخ


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

| دور حلقه | مقدار <code dir="ltr">planet</code> | خروجی |
|:---:|:---:|:---|
| ۱ | <code dir="ltr">"عطارد"</code> | 🪐 عطارد |
| ۲ | <code dir="ltr">"زهره"</code> | 🪐 زهره |
| ۳ | <code dir="ltr">"زمین"</code> | 🪐 زمین |
| ۴ | <code dir="ltr">"مریخ"</code> | 🪐 مریخ |
| — | اعضای لیست تمام شد | حلقه پایان می‌یابد |

تعداد دورهای حلقه **دقیقاً برابر** با تعداد اعضای لیست است.

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #fce4ec; color: #1b1b1b; border-right: 6px solid #c2185b; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 🧠 مثال علوم اعصاب: تبدیل واحد زمان واکنش

در آزمایش‌های **زمان واکنش** (reaction time)، شرکت‌کننده باید به محض دیدن یک نور روی صفحه، دکمه‌ای را فشار دهد. نرم‌افزار، زمان واکنش را بر حسب **میلی‌ثانیه** ثبت می‌کند. بیا آن‌ها را به **ثانیه** تبدیل کنیم (هر ثانیه ۱۰۰۰ میلی‌ثانیه است).

</div>

In [4]:
reaction_times_ms = [312, 287, 345, 298, 401]

for rt in reaction_times_ms:
    seconds = rt / 1000
    print(rt, "میلی‌ثانیه =", seconds, "ثانیه")

312 میلی‌ثانیه = 0.312 ثانیه
287 میلی‌ثانیه = 0.287 ثانیه
345 میلی‌ثانیه = 0.345 ثانیه
298 میلی‌ثانیه = 0.298 ثانیه
401 میلی‌ثانیه = 0.401 ثانیه


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

## ۴. الگوی جمع‌کننده (Accumulator)

یکی از مهم‌ترین کاربردهای حلقه، **جمع کردن** اعضای لیست است. ایده ساده است:

1. **پیش از حلقه** یک متغیر به نام <code dir="ltr">total</code> می‌سازیم و مقدارش را <code dir="ltr">0</code> می‌گذاریم.
2. **در هر دور**، عضو فعلی را به <code dir="ltr">total</code> اضافه می‌کنیم.
3. **بعد از حلقه**، <code dir="ltr">total</code> مجموع همه‌ی اعضاست.

</div>

In [5]:
reaction_times_ms = [312, 287, 345, 298, 401]

total = 0
for rt in reaction_times_ms:
    total = total + rt
    print("عدد فعلی:", rt, "  ← مجموع تا اینجا:", total)

average = total / len(reaction_times_ms)
print("--------------------")
print("مجموع:", total)
print("میانگین زمان واکنش:", average, "میلی‌ثانیه")

عدد فعلی: 312   ← مجموع تا اینجا: 312
عدد فعلی: 287   ← مجموع تا اینجا: 599
عدد فعلی: 345   ← مجموع تا اینجا: 944
عدد فعلی: 298   ← مجموع تا اینجا: 1242
عدد فعلی: 401   ← مجموع تا اینجا: 1643
--------------------
مجموع: 1643
میانگین زمان واکنش: 328.6 میلی‌ثانیه


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #e8f5e9; color: #1b1b1b; border-right: 6px solid #2e7d32; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 💡 خط <code dir="ltr">total = total + rt</code> را چطور بخوانیم؟

در ریاضی، <code dir="ltr">total = total + rt</code> بی‌معنی است! اما در برنامه‌نویسی، علامت <code dir="ltr">=</code> یعنی **«مقدار سمت راست را حساب کن و در متغیر سمت چپ بگذار»**.

پس این خط می‌گوید: «مقدار فعلی <code dir="ltr">total</code> را بردار، <code dir="ltr">rt</code> را به آن اضافه کن و نتیجه را دوباره در <code dir="ltr">total</code> بگذار». در جلسه‌ی ۶ شکل کوتاه‌تر آن یعنی <code dir="ltr">total += rt</code> را هم یاد می‌گیریم.

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #e3f2fd; color: #1b1b1b; border-right: 6px solid #1565c0; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 💻 مثال علوم کامپیوتر: وزن همینگ

در علوم کامپیوتر، **تعداد بیت‌های ۱** در یک دنباله‌ی بیت را **وزن همینگ** (Hamming weight) می‌نامند. این عدد در رمزنگاری و تشخیص خطا کاربرد دارد. (درباره‌ی ریچارد همینگ در جلسه‌ی ۸ بیشتر می‌خوانیم.)

چون بیت‌ها فقط <code dir="ltr">0</code> یا <code dir="ltr">1</code> هستند، کافی است همه را با هم جمع کنیم!

</div>

In [6]:
bits = [1, 0, 1, 1, 0, 0, 1, 0]

ones = 0
for bit in bits:
    ones = ones + bit

print("وزن همینگ این بایت:", ones)

وزن همینگ این بایت: 4


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

## ۵. الگوی شمارنده

شمارنده شبیه جمع‌کننده است؛ با این تفاوت که در هر دور، به جای خود عضو، **عدد ۱** را اضافه می‌کنیم. این کار وقتی معنا پیدا می‌کند که بخواهیم چیزی را بشماریم.

</div>

In [7]:
# 🎵 تعداد نت‌های یک ملودی
melody = ["می", "ر", "دو", "ر", "می", "می", "می"]

count = 0
for note in melody:
    count = count + 1

print("این ملودی", count, "نت دارد.")
print("بررسی با len:", len(melody))

این ملودی 7 نت دارد.
بررسی با len: 7


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

البته برای شمردن همه‌ی اعضا، <code dir="ltr">len</code> کافی است. شمارنده وقتی واقعاً به کار می‌آید که بخواهیم فقط عضوهایی با **شرط خاص** را بشماریم (مثلاً فقط نت‌های «می»). این را در جلسه‌ی ۵ با <code dir="ltr">if</code> یاد می‌گیریم.

## ۶. ساختن متن در حلقه

می‌توانیم با جمع کردن رشته‌ها (<code dir="ltr">+</code>) در حلقه، یک متن بلندتر بسازیم:

</div>

In [8]:
# 🧬 بازهای یک رشته‌ی DNA را پشت سر هم می‌چسبانیم و با خط تیره جدا می‌کنیم.
dna = ["A", "T", "G", "C", "G", "A"]

sequence = ""
for base in dna:
    sequence = sequence + base + "-"

print(sequence)

A-T-G-C-G-A-


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

## ۷. حلقه روی یک رشته

حلقه‌ی <code dir="ltr">for</code> فقط مخصوص لیست نیست. یک **رشته** هم دنباله‌ای از حروف است و می‌توانیم روی حروفش حلقه بزنیم:

</div>

In [9]:
word = "NEURON"

for letter in word:
    print(letter, "→", letter.lower())

N → n
E → e
U → u
R → r
O → o
N → n


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #f1f8e9; color: #1b1b1b; border-right: 6px solid #558b2f; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 🔬 مثال فیزیک و محیط زیست: تبدیل دما

دمای هوای یک هفته بر حسب **سلسیوس** ثبت شده است. فرمول تبدیل به **فارنهایت** این است:

$$F = C \times 1.8 + 32$$

</div>

In [10]:
temps_c = [18, 21, 25, 30, 27, 22, 19]

for c in temps_c:
    f = c * 1.8 + 32
    print(c, "°C  =", round(f, 1), "°F")

18 °C  = 64.4 °F
21 °C  = 69.8 °F
25 °C  = 77.0 °F
30 °C  = 86.0 °F
27 °C  = 80.6 °F
22 °C  = 71.6 °F
19 °C  = 66.2 °F


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #fff3e0; color: #1b1b1b; border-right: 6px solid #ef6c00; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### ⚠️ خطاهای رایج در حلقه‌ی for

1. **فراموش کردن دونقطه:** <code dir="ltr">for x in nums</code> ← خطای <code dir="ltr">SyntaxError</code>
2. **فراموش کردن تورفتگی:** اگر بدنه‌ی حلقه تورفتگی نداشته باشد، خطای <code dir="ltr">IndentationError</code> می‌گیری.
3. **تورفتگی اشتباه:** اگر <code dir="ltr">print</code> نتیجه‌ی نهایی را داخل حلقه بگذاری، به جای یک بار، در هر دور چاپ می‌شود.
4. **تغییر متغیر حلقه، لیست را عوض نمی‌کند:** سلول زیر را ببین.

</div>

In [11]:
numbers = [1, 2, 3]

for n in numbers:
    n = n * 10   # فقط متغیر n عوض می‌شود، نه خود لیست!

print(numbers)   # هنوز [1, 2, 3] است

[1, 2, 3]


<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #f3e5f5; color: #1b1b1b; border-right: 6px solid #7b1fa2; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 📜 داستان تاریخ: از خوارزمی تا تورینگ

**محمد بن موسی خوارزمی** (حدود ۷۸۰ تا ۸۵۰ میلادی)، دانشمند ایرانی اهل **خوارزم**، در «بیت‌الحکمه»ی بغداد کار می‌کرد. او کتاب‌هایی درباره‌ی حساب با ارقام هندی و حل معادله‌ها نوشت. وقتی کتاب‌هایش به لاتین ترجمه شد، نامش به صورت **Algoritmi** نوشته شد و واژه‌ی **الگوریتم** (algorithm) از همین‌جا آمد. واژه‌ی **جبر** (algebra) هم از نام کتاب او، «الجبر و المقابله»، گرفته شده است.

الگوریتم یعنی **دستورالعملی گام‌به‌گام برای حل یک مسئله**؛ و حلقه یکی از مهم‌ترین ابزارهای هر الگوریتم است.

بیش از هزار سال بعد، ریاضی‌دان انگلیسی **آلن تورینگ** (Alan Turing، ۱۹۱۲ تا ۱۹۵۴) در سال ۱۹۳۶ ماشینی خیالی را تعریف کرد که امروز **ماشین تورینگ** نام دارد. این ماشین یک **نوار** بلند دارد که به خانه‌های پشت سر هم تقسیم شده و در هر خانه یک نماد نوشته شده است. ماشین **خانه‌به‌خانه** جلو می‌رود، نماد را می‌خواند و طبق قوانینش تصمیم می‌گیرد؛ درست شبیه حلقه‌ای که روی اعضای یک لیست حرکت می‌کند! 🎞️

تورینگ در جنگ جهانی دوم در **بلچلی پارک** به شکستن رمز ماشین **انیگما**ی آلمان کمک کرد. در سال ۱۹۵۰ هم مقاله‌ای نوشت که «**آزمون تورینگ**» را برای سنجش هوش ماشین پیشنهاد می‌داد. معتبرترین جایزه‌ی علوم کامپیوتر، **جایزه‌ی تورینگ**، به نام اوست.

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #e0f2f1; color: #1b1b1b; border-right: 6px solid #00695c; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 🧩 کوییز سریع: خروجی را حدس بزن!

قبل از اجرا، روی کاغذ بنویس این برنامه چه چیزی چاپ می‌کند. حواست به تورفتگی خط آخر باشد!

</div>

In [ ]:
scores = [4, 7, 1]
total = 0
for s in scores:
    total = total + s
    print("در حلقه:", total)
print("پایان:", total)

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

<details>
<summary><b>👀 دیدن پاسخ</b></summary>

<div dir="ltr" align="left">

```
در حلقه: 4
در حلقه: 11
در حلقه: 12
پایان: 12
```

</div>

</details>

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #fffde7; color: #1b1b1b; border-right: 6px solid #f9a825; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### ✏️ تمرین ۱ (آسان): سلام به همه

لیستی از نام **چهار دوستت** بساز. با حلقه‌ی <code dir="ltr">for</code> به هر کدام سلام کن؛ مثلاً: <code dir="ltr">سلام رضا! 👋</code>

</div>

In [13]:
# کد خودت را اینجا بنویس ✍️

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

<details>
<summary><b>👀 دیدن پاسخ پیشنهادی</b></summary>

<div dir="ltr" align="left">

```python
friends = ["رضا", "الناز", "کیان", "ترانه"]
for friend in friends:
    print("سلام", friend + "! 👋")
```

</div>

</details>

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #fffde7; color: #1b1b1b; border-right: 6px solid #f9a825; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### ✏️ تمرین ۲ (آسان): مربع‌ها

برای هر عدد در لیست <code dir="ltr">[1, 2, 3, 4, 5]</code>، خود عدد و **مربع** آن را چاپ کن. (مربع یعنی عدد ضرب در خودش.)

</div>

In [14]:
# کد خودت را اینجا بنویس ✍️

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

<details>
<summary><b>👀 دیدن پاسخ پیشنهادی</b></summary>

<div dir="ltr" align="left">

```python
for n in [1, 2, 3, 4, 5]:
    print(n, "→", n * n)
```

</div>

</details>

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #fffde7; color: #1b1b1b; border-right: 6px solid #f9a825; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### ✏️ تمرین ۳ (متوسط): میانگین نرخ شلیک نورون

🧠 نورون‌ها با فرستادن پالس‌های الکتریکی کوتاهی به نام **اسپایک** (spike) با هم حرف می‌زنند. تعداد اسپایک‌ها در هر ثانیه را **نرخ شلیک** (firing rate) می‌نامند و واحد آن **هرتز** (Hz) است.

لیست زیر نرخ شلیک یک نورون را در ۶ آزمایش نشان می‌دهد. با الگوی جمع‌کننده، **مجموع** و **میانگین** نرخ شلیک را حساب و چاپ کن. (از تابع آماده‌ی <code dir="ltr">sum</code> استفاده نکن!)

</div>

In [15]:
firing_rates = [12, 18, 15, 22, 9, 14]
# کد خودت را اینجا بنویس ✍️

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

<details>
<summary><b>👀 دیدن پاسخ پیشنهادی</b></summary>

<div dir="ltr" align="left">

```python
firing_rates = [12, 18, 15, 22, 9, 14]
total = 0
for rate in firing_rates:
    total = total + rate
average = total / len(firing_rates)
print("مجموع:", total)
print("میانگین نرخ شلیک:", average, "هرتز")
```

</div>

</details>

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #fffde7; color: #1b1b1b; border-right: 6px solid #f9a825; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### ✏️ تمرین ۴ (متوسط): ضرب‌کننده

الگوی جمع‌کننده را تغییر بده تا **حاصل‌ضرب** همه‌ی اعداد لیست <code dir="ltr">[2, 3, 4, 5]</code> را حساب کند.

💡 **راهنمایی:** مقدار اولیه‌ی متغیر را چه بگذاری؟ اگر <code dir="ltr">0</code> بگذاری چه می‌شود؟

</div>

In [16]:
# کد خودت را اینجا بنویس ✍️

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

<details>
<summary><b>👀 دیدن پاسخ پیشنهادی</b></summary>

<div dir="ltr" align="left">

```python
product = 1   # اگر 0 بگذاریم، حاصل‌ضرب همیشه 0 می‌شود!
for n in [2, 3, 4, 5]:
    product = product * n
print(product)   # 120
```

</div>

</details>

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #fffde7; color: #1b1b1b; border-right: 6px solid #f9a825; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### ✏️ تمرین ۵ (چالشی): برعکس کردن کلمه با حلقه

بدون استفاده از برش <code dir="ltr">[::-1]</code>، با یک حلقه‌ی <code dir="ltr">for</code> کلمه‌ی <code dir="ltr">"PYTHON"</code> را برعکس کن.

💡 **راهنمایی:** هر حرف جدید را به **اول** متن قبلی بچسبان، نه به آخرش: <code dir="ltr">reversed_word = letter + reversed_word</code>

یک جدول ردیابی هم روی کاغذ بکش تا ببینی در هر دور، مقدار <code dir="ltr">reversed_word</code> چه می‌شود.

</div>

In [17]:
# کد خودت را اینجا بنویس ✍️

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px;">

<details>
<summary><b>👀 دیدن پاسخ پیشنهادی</b></summary>

<div dir="ltr" align="left">

```python
word = "PYTHON"
reversed_word = ""
for letter in word:
    reversed_word = letter + reversed_word
print(reversed_word)   # NOHTYP
```

</div>

</details>

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #eceff1; color: #1b1b1b; border-right: 6px solid #455a64; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 📝 خلاصه‌ی جلسه

- حلقه‌ی <code dir="ltr">for item in my_list:</code> برای **هر عضو** لیست یک بار اجرا می‌شود.
- **دونقطه** و **تورفتگی** تعیین می‌کنند کدام خط‌ها داخل حلقه هستند.
- **الگوی جمع‌کننده:** <code dir="ltr">total = 0</code> پیش از حلقه، <code dir="ltr">total = total + x</code> داخل حلقه.
- **الگوی شمارنده:** <code dir="ltr">count = 0</code> پیش از حلقه، <code dir="ltr">count = count + 1</code> داخل حلقه.
- روی **رشته** هم می‌توان حلقه زد: <code dir="ltr">for letter in "ABC":</code>
- تغییر دادن متغیر حلقه، خود لیست را تغییر **نمی‌دهد**.

</div>

<div dir="rtl" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; line-height: 2; font-size: 15px; background: #fbe9e7; color: #1b1b1b; border-right: 6px solid #d84315; border-radius: 12px; padding: 14px 20px; margin: 12px 0;">

### 🏠 تکلیف خانه

1. لیستی از **قیمت ۵ کتاب** بساز و با حلقه، مجموع قیمت‌ها را حساب کن. اگر کتاب‌فروشی ۱۰ درصد تخفیف بدهد، مبلغ نهایی چقدر است؟
2. 🧠 لیست <code dir="ltr">[0.31, 0.29, 0.35, 0.40]</code> زمان واکنش را بر حسب **ثانیه** نشان می‌دهد. با حلقه آن‌ها را به **میلی‌ثانیه** تبدیل و چاپ کن.
3. برای کلمه‌ی <code dir="ltr">"INFORMATION"</code>، با حلقه هر حرف را در یک خط جدا چاپ کن و در پایان تعداد حروف را با الگوی شمارنده نشان بده.

</div>

<div dir="rtl" align="center" style="font-family: Vazirmatn, Vazir, Tahoma, 'Segoe UI', sans-serif; background: #263238; color: #eceff1; padding: 18px; border-radius: 14px; margin-top: 16px;">

**پایان جلسه‌ی ۲** ✨

در جلسه‌ی بعد سراغ **متدهای لیست** می‌رویم. 👋

</div>